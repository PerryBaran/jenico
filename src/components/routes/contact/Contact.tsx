import {
  useReducer,
  useRef,
  FormEvent,
  ChangeEvent,
} from "react";
import Background from "../../background/Background";
import style from "./contact.module.css";
import sendEmail from "../../../services/emailJS";
import { background } from "../../../media/images/index";

const initialValues = {
  name: "",
  email: "",
  subject: "",
  message: "",
};

type Form = typeof initialValues;

type ReducerAction =
  | { type: "reset" }
  | { type: "update"; key: string; value: string };

const reducer = (state: Form, action: ReducerAction) => {
  switch (action.type) {
    case "update":
      return {
        ...state,
        [action.key]: action.value,
      };
    case "reset":
      return initialValues;
    default:
      throw new Error("invalid action type");
  }
};

function Contact(props: { handleFormFocus: (value: boolean) => void}) {
  const { handleFormFocus } = props;
  const [state, dispatch] = useReducer(reducer, initialValues);
  const formRef = useRef<HTMLFormElement>(null);

  const handleChange = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    dispatch({
      type: "update",
      key: e.target.name,
      value: e.target.value,
    });
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!state.name || !state.email || !state.subject || !state.message) {
      alert("all form fields must be filled");
      return;
    };
    if (!state.email.match(/[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$/)) {
      alert("invalid email");
      return;
    };
    sendEmail(formRef.current);
    dispatch({ type: "reset" });
  };

  const handleFocus = () => {
    handleFormFocus(true);
  };

  const handleBlur = () => {
    handleFormFocus(false);
  };

  return (
    <>
      <Background src={background} />
      <section className={style.container}>
        <form ref={formRef} onSubmit={handleSubmit}>
          <div>
            <label htmlFor="name">Name</label>
            <input
              name="name"
              required
              value={state.name}
              onChange={handleChange}
              onFocus={handleFocus}
              onBlur={handleBlur}
            />
          </div>
          <div>
            <label htmlFor="email">Email</label>
            <input
              type="email"
              name="email"
              required
              pattern="[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$"
              value={state.email}
              onChange={handleChange}
              onFocus={handleFocus}
              onBlur={handleBlur}
            />
          </div>
          <div>
            <label htmlFor="subject">Subject</label>
            <input
              name="subject"
              required
              value={state.subject}
              onChange={handleChange}
              onFocus={handleFocus}
              onBlur={handleBlur}
            />
          </div>
          <div>
            <label htmlFor="message">Message</label>
            <textarea
              name="message"
              required
              value={state.message}
              onChange={handleChange}
              onFocus={handleFocus}
              onBlur={handleBlur}
            />
          </div>
          <div>
            <button type="submit" name="submit">
              Submit
            </button>
          </div>
        </form>
      </section>
    </>
  );
}

export default Contact;
