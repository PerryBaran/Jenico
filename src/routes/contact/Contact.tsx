import React, { useState, useRef, FormEvent } from 'react';
import style from './contact.module.css';
import emailjs from 'emailjs-com'

function Contact() {
    const [name, setName] = useState('');
    const [email, setEmail] = useState('');
    const [subject, setSubject] = useState('');
    const [message, setMessage] = useState('');


    const form = useRef<HTMLFormElement|null>(null);

    const sendEmail = (e:FormEvent<HTMLFormElement>) => {
      e.preventDefault();

      if (form.current) {
        emailjs.sendForm('service_szxchwk', 'template_ldjw82j', form.current, 'kavY_LmpokDn2OKmV')
        .then((result) => {
            alert('Email sent successfully')
        }, (error) => {
            alert(`Oops, something went wrong: ${error.text}`);
        });
      };

      setName('');
      setEmail('');
      setSubject('');
      setMessage('');
    };

    return (
        <section className={style.container}>
            <form ref={form} onSubmit={e => sendEmail(e)}>
                <label htmlFor='name'>Name</label>
                <input 
                    type='text' 
                    name='name' 
                    required
                    value={name}
                    onChange={e => setName(e.target.value)}/>
                <label htmlFor='email'>Email </label>
                <input 
                    type='email' 
                    name='email' 
                    required
                    value={email}
                    onChange={e => setEmail(e.target.value)}/>
                <label htmlFor='subject'>Subject</label>
                <input 
                    type='text' 
                    name='subject' 
                    required
                    value={subject}
                    onChange={e => setSubject(e.target.value)}/>
                <label htmlFor='message'>Message</label>
                <textarea 
                    name='message' 
                    required
                    value={message}
                    onChange={e => setMessage(e.target.value)}/>
                <button type='submit' name='submit'>Submit</button>
            </form>
        </section>
    );
};

export default Contact;