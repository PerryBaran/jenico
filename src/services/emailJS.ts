import emailjs from "emailjs-com";

const sendEmail = async (content: HTMLFormElement | null) => {
  if (content) {
    await emailjs
      .sendForm(
        "service_szxchwk",
        "template_ldjw82j",
        content,
        "kavY_LmpokDn2OKmV"
      )
      .then(
        () => {
          alert("Email sent successfully");
        },
        (error) => {
          alert(`Oops, something went wrong: ${error.text}`);
        }
      );
  }
};

export default sendEmail;
