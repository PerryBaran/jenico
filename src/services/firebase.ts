import { initializeApp } from "firebase/app";
import { getStorage, ref, getDownloadURL } from "firebase/storage";

const firebaseConfig = {
    apiKey: "AIzaSyDzb4R64c1c8GJIbf8aRDXzBOB-lSvSdss",
    authDomain: "jenico-b1936.firebaseapp.com",
    projectId: "jenico-b1936",
    storageBucket: "jenico-b1936.appspot.com",
    messagingSenderId: "873325094087",
    appId: "1:873325094087:web:a521e37d3b7dadbb28cae2"
};

initializeApp(firebaseConfig);
const storage = getStorage();

const getURL = async (path: string) => {
  try {
    const fileRef = ref(storage, path);
    const url = await getDownloadURL(fileRef);
    return url
  } catch(err) {
    if (err instanceof Error) {
      throw new Error(`Failed to retrieve song URL ${err.message}`);
    }
    throw new Error('Failed to retrieve song URL');
  }
};

export { getURL };