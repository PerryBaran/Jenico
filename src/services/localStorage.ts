export function getLocalStorage(value:string) {
    if (!localStorage.getItem(value)){
        return undefined;
    } else {
        const retrieveStorage = localStorage.getItem(value);
        if (retrieveStorage) {
            return JSON.parse(retrieveStorage)
        }
        return undefined;
    }
}

export function populateStorage(name: string, value: any) {
    localStorage.setItem(name, JSON.stringify(value));
}