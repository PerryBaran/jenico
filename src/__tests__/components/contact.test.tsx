import {
  render,
  screen,
  cleanup,
  fireEvent,
  act,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import Contact from "../../components/routes/contact/Contact";
import * as sendEmail from "../../services/emailJS";

describe("Contact", () => {
  const validProps = {
    handleFormFocus: jest.fn(),
  };

  afterEach(cleanup);

  test("snapshot", () => {
    const { asFragment } = render(
      <Contact handleFormFocus={validProps.handleFormFocus} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("content", () => {
    render(<Contact handleFormFocus={validProps.handleFormFocus} />);
    const inputs = screen.getAllByRole("textbox");

    expect(screen.getByRole("img")).toBeInTheDocument();
    expect(screen.getByText(/name/i)).toHaveAttribute("for", "name");
    expect(screen.getByText(/email/i)).toHaveAttribute("for", "email");
    expect(screen.getByText(/subject/i)).toHaveAttribute("for", "subject");
    expect(screen.getByText(/message/i)).toHaveAttribute("for", "message");
    expect(screen.getByText(/submit/i)).toHaveAttribute("type", "submit");
    expect(inputs[0]).toHaveAttribute("name", "name");
    expect(inputs[0]).toHaveAttribute("required");
    expect(inputs[1]).toHaveAttribute("type", "email");
    expect(inputs[1]).toHaveAttribute("name", "email");
    expect(inputs[1]).toHaveAttribute("required");
    expect(inputs[2]).toHaveAttribute("name", "subject");
    expect(inputs[2]).toHaveAttribute("required");
    expect(inputs[3]).toHaveAttribute("name", "message");
    expect(inputs[3]).toHaveAttribute("required");
  });

  test("focus and blur", () => {
    render(<Contact handleFormFocus={validProps.handleFormFocus} />);
    const inputs = screen.getAllByRole("textbox");

    expect(validProps.handleFormFocus).toBeCalledTimes(0);
    inputs.forEach((input) => {
      input.focus();
      expect(validProps.handleFormFocus).lastCalledWith(true);
      input.blur();
      expect(validProps.handleFormFocus).lastCalledWith(false);
    });
  });

  test("form submit", () => {
    const mockSendEmail = jest.fn();
    const mockAlert = jest.fn();
    jest.spyOn(sendEmail, "default").mockImplementation(mockSendEmail);
    jest.spyOn(window, "alert").mockImplementation(mockAlert);
    render(<Contact handleFormFocus={validProps.handleFormFocus} />);
    const inputs = screen.getAllByRole("textbox");
    const submit = screen.getByText(/submit/i);

    submit.click();
    expect(mockSendEmail).toBeCalledTimes(0);
    fireEvent.change(inputs[0], { target: { value: "valid name" } });
    submit.click();
    expect(mockAlert).lastCalledWith("all form fields must be filled");
    expect(mockSendEmail).toBeCalledTimes(0);
    fireEvent.change(inputs[1], { target: { value: "not valid email" } });
    submit.click();
    expect(mockAlert).lastCalledWith("all form fields must be filled");
    expect(mockSendEmail).toBeCalledTimes(0);
    fireEvent.change(inputs[2], { target: { value: "valid subject" } });
    expect(mockAlert).lastCalledWith("all form fields must be filled");
    submit.click();
    expect(mockSendEmail).toBeCalledTimes(0);
    fireEvent.change(inputs[3], { target: { value: "valid@email.com" } });
    submit.click();
    expect(mockAlert).lastCalledWith("invalid email");
    fireEvent.change(inputs[1], { target: { value: "valid@email.com" } });
    act(() => submit.click());
    expect(mockSendEmail).toBeCalledTimes(1);
  });
});
