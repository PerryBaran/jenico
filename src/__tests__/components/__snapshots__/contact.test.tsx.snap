// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="background"
      src="background.jpg"
    />
  </div>
  <section
    class="container"
  >
    <form>
      <div>
        <label
          for="name"
        >
          Name
        </label>
        <input
          name="name"
          required=""
          value=""
        />
      </div>
      <div>
        <label
          for="email"
        >
          Email
        </label>
        <input
          name="email"
          pattern="[a-z0-9._%+-]+@[a-z0-9.-]+\\\\.[a-z]{2,4}$"
          required=""
          type="email"
          value=""
        />
      </div>
      <div>
        <label
          for="subject"
        >
          Subject
        </label>
        <input
          name="subject"
          required=""
          value=""
        />
      </div>
      <div>
        <label
          for="message"
        >
          Message
        </label>
        <textarea
          name="message"
          required=""
        />
      </div>
      <div>
        <button
          name="submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
